Fix TypeError when changing click behavior to drill-through menu & default click behavior does not have selected state

ibedwi opened the pull request:
Closes https://github.com/metabase/metabase/issues/35354 , https://github.com/metabase/metabase/issues/34848

### Description

Uncaught TypeError: Cannot read properties of undefined (reading 'type')

#### Root cause

As @kamilmielnik mentioned:
> nextClickBehavior is undefined in handleChangeSettings and [this throws](https://github.com/metabase/metabase/blob/8f8ab297744d7188101aef15b758f678af36aa6f/frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/ClickBehaviorSidebar.tsx#L112).

([Here's where the value sent from](https://github.com/metabase/metabase/blob/45315fa445f087cbc216d6e935555028e28032d8/frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/TypeSelector/TypeSelector.tsx#L75)) 

Another issue is in the UI, the drill through option is not selected:


![CleanShot 2023-11-16 at 20 02 38@2x](https://github.com/metabase/metabase/assets/15979404/896c5b71-7fe4-4a71-9693-a59b7849c575)



while the other option is in selected state when it selected:

![CleanShot 2023-11-16 at 20 04 42@2x](https://github.com/metabase/metabase/assets/15979404/c63795a4-9759-4937-b7c7-23e1a6da1b02)


#### How this fixed

1. Only check for `nextClickBehavior.type` when it's not undefined
2. In the `TypeSelector` component, the selected state is adjusted with the drill-through menu value

### How to verify

Describe the steps to verify that the changes are working as expected.

1. Create a dashboard and add a line chart question to it
2. Edit the dashboard
3. Edit question's custom click behavior
4. Click "Open the Metabase drill-through menu"

You shouldn't see any JS error message and the drill-through menu should be selected

### Demo

https://github.com/metabase/metabase/assets/15979404/3b480a8b-d75e-4b87-8a67-aa39ba49df58

#### E2E

https://github.com/metabase/metabase/assets/15979404/7cc5b7fc-7666-41da-8986-cf9635686ca6

<img width="1800" alt="CleanShot 2023-12-05 at 22 53 15@2x" src="https://github.com/metabase/metabase/assets/15979404/bfd302c2-ed4b-4864-b0bd-f947c57ce788">

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

ibedwi committed:
fix: type error on click behavior change

ibedwi committed:
chore: add comments

ibedwi committed:
chore: tidy up comments

ibedwi committed:
chore: reset changes

kamilmielnik commented:
> Another issue is in the UI, the drill through option is not selected:

This issue has been reported separately: #34848.
@ibedwi can you please update the PR description?

----

As for the PR, can you please add some assertions/tests for the fixed issues?
We have an existing test suite for the click behavior in `e2e/test/scenarios/dashboard-cards/click-behavior.cy.spec.js`.

ibedwi committed:
add aria-selected & selected drill-through menu assertion

ibedwi committed:
chore: update assertion comment

ibedwi commented on e2e/test/scenarios/dashboard-cards/click-behavior.cy.spec.js:
```diff
@@ -195,6 +195,15 @@ describe("scenarios > dashboard > dashboard cards > click behavior", () => {
 
       getDashboardCard().realHover().icon("click").click();
+
+      // When the default menu is selected, it should've visual cue (metabase#34848)
+      cy.get("aside")
+        .findByText("Open the Metabase drill-through menu")
+        .parent()
+        .parent()
+        .should("have.attr", "aria-selected", "true")
+        .should("have.css", "background-color", "rgb(80, 158, 227)");
```
I added the aria-selected attribute and included it in the test. I've also tested whether the menu has the expected background-color for the selected state. Do you have any input on this matter? @kamilmielnik

kamilmielnik commented on frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/TypeSelector/TypeSelector.tsx:
```diff
@@ -33,6 +33,7 @@ export const BehaviorOption = ({
     isSelected={selected}
     onClick={onClick}
     disabled={disabled}
+    aria-selected={selected}
```
Good call on adding `aria-selected` :+1: 
However I believe we should add this attribute inside of `SidebarItem.Selectable` instead - this way we'd handle all `SidebarItem.Selectable` usages at once.

ibedwi committed:
refactor: move aria-selected inside selectable item

ibedwi committed:
remove 'menu' from click behavior option

ibedwi committed:
test: add e2e test fro 35354

ibedwi committed:
chore: clean up demo and update assertion

ibedwi commented:
Hi @kamilmielnik , I apologize for taking too long to update.

I get rid of `"menu"` option as [you suggested here](https://github.com/metabase/metabase/pull/35863#discussion_r1403079700).

I'm not sure how to do the e2e test. I ended up creating a test scenario that [expected the specific error message not to be thrown in the console](https://github.com/metabase/metabase/pull/35863/files#diff-d20bdd718e35d79993c0797bb9bc9df3a74f2f54b87b55196ba26809eeddb4d7R76).

Here's how the success assertion looks like:

https://github.com/metabase/metabase/assets/15979404/e1f83728-1100-4394-850e-24433be8bfa3

And here's how the failed assertion looks like:

https://github.com/metabase/metabase/assets/15979404/c5466338-ce0b-4a47-a7d6-71da30e847f7

Would love to hear your feedback. Thanks!

ibedwi committed:
refactor: move assertion to existing scenario

ibedwi committed:
chore: add log on e2e test

ibedwi commented on e2e/test/scenarios/dashboard-cards/reproductions/35354-click-behavior-setting-custom-behavior-type-error.cy.spec.js:
```diff
@@ -0,0 +1,146 @@
+
+const URL = "https://metabase.com/";
+
+describe("scenarios > dashboard > dashboard cards > click behavior", () => {
```
Updated as you requested. 

<img width="1800" alt="CleanShot 2023-12-05 at 22 53 15@2x" src="https://github.com/metabase/metabase/assets/15979404/7d908253-3876-4e28-b908-26e941918a10">

I got your point and realized that dedicated e2e test suite was an overkill. Moreover, what's really different is just a few lines that only assert the error. By placing the assertion [`testChangingBackToDefaultBehavior`](https://github.com/metabase/metabase/blob/1d7209d/e2e/test/scenarios/dashboard-cards/click-behavior.cy.spec.js#L1602), it will be called whenever `testChangingBackToDefaultBehavior` and the assertion will be checked each time the default setting is clicked. Thank you for your feedback! 🙏

kamilmielnik reviewed (dismissed):
LGTM :+1:

ibedwi committed:
feat: update name for actionMenu

ibedwi committed:
fix: unknown type in 34845

ibedwi commented on frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/ClickBehaviorSidebar.tsx:
```diff
@@ -158,6 +160,7 @@ export function ClickBehaviorSidebar({
     }
     setOriginalColumnVizSettings(null);
     setSelectedColumn(null);
+    setTypeSelectorVisible(true);
```
Hi @kamilmielnik . This issue https://github.com/metabase/metabase/issues/34845 has two problems:

1. The current selected type was "Unknown".
2. The sidebar should show type selector instead of just one option.

The first issue solved by the `changedType` fixes above. The second issue solved by this line. When the user click the header, the type selector visibility isn't changed.

I've remove the TODO related to the issue above and rerun the E2E tests for the click behavior, here's what I got locally:

![CleanShot 2023-12-14 at 01 25 50@2x](https://github.com/metabase/metabase/assets/15979404/4cc1d9f7-1649-4c1e-957e-f13f98b61322)

Would love to hear your feedback!

kamilmielnik replied:
@ibedwi I tried it with and without this line and I didn't notice any difference.
Could you please write down the steps where it makes a difference?

kamilmielnik replied:
Thank you for the detailed explanation!

> By updating the isTypeSelectorVisible value to true when the user taken back to step 1 will make sure that the **user will see the type selection step again when clicking on any column**.

This behavior is only desired when editing the default "drill-through" click behavior.
:x: When editing "link" or "crossfilter" click behaviors, user should be taken to the very last step of the wizard, not the click behavior type selection.

----

To rephrase, just for clarity:

1. When user selects a column that has the default "drill-through" click behavior - they should be taken [here](https://github.com/metabase/metabase/assets/6830683/a8220771-b9d2-4845-9e25-2b871898f122).
2. When user selects a column that has either "link" or "crossfilter" click behavior - they should be taken [here](https://github.com/metabase/metabase/assets/6830683/9785b135-d23d-4b58-86d1-fd41ec2b8ca1).

----

And I think you missed my [other comment](https://github.com/metabase/metabase/pull/35863#issuecomment-1844978101) - the e2e tests need to be updated.

ibedwi replied:
> To rephrase, just for clarity:
> 
> When user selects a column that has the default "drill-through" click behavior - they should be taken [here](https://github.com/metabase/metabase/assets/6830683/a8220771-b9d2-4845-9e25-2b871898f122).
When user selects a column that has either "link" or "crossfilter" click behavior - they should be taken [here](https://github.com/metabase/metabase/assets/6830683/9785b135-d23d-4b58-86d1-fd41ec2b8ca1).

Hi @kamilmielnik , 
I apologize for my misunderstanding about the behavior. I revert the changes that  always turn `isTypeSelectorVisible` to true.

To keep showing the default ("drill-through") click behavior, I change [this line](https://github.com/metabase/metabase/pull/35863/files#diff-7111f88ba09a9ef63ac313c293e95c130e0b4c260e3eb824975fb6c43c8b9b07R69) to:

```tsx
// ClickBehaviorSidebarContent.tsx

if (isTypeSelectorVisible || finalClickBehavior.type === "actionMenu") //...
```

I think this is a correct way to show the "drill-through" click behavior since it represented by `undefined` or `ActionMenuClickBehavior` type. Since we already create a variable that represent the click behavior, `finalClickBehavior`, I just use that variable in the conditional statement.

This is the demo of the latest behavior:


https://github.com/metabase/metabase/assets/15979404/98573446-5c1f-4579-b8ed-becfc0f06590

-- 


> And I think you missed my https://github.com/metabase/metabase/pull/35863#issuecomment-1844978101 - the e2e tests need to be updated.

I've also updated the e2e tests. I think I missed something and the changes were reverted when I rebased this PR from the `master` branch.

kamilmielnik reviewed (commented):
Hey @ibedwi!

Apologies for late response, I was OOO for Christmas break.

- There are some conflicts - can you please merge the latest `master` in?
- It seems that there is a TS error: https://github.com/metabase/metabase/actions/runs/7199526117/job/20089367211?pr=35863
    ```
    Error: frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/TypeSelector/TypeSelector.tsx(39,57): error 
    TS2345: Argument of type '"menu" | ClickBehaviorType' is not assignable to parameter of type 'ClickBehaviorType'.
      Type '"menu"' is not assignable to type 'ClickBehaviorType'.
    ```

ibedwi committed:
fix: test after rebase

ibedwi committed:
fix: remove unused union type "menu"

ibedwi committed:
fix: use master version of click-behavior

ibedwi committed:
fix: readd the tests

ibedwi commented:
> Hey @ibedwi!
> 
> Apologies for late response, I was OOO for Christmas break.
> 
> * There are some conflicts - can you please merge the latest `master` in?
> * It seems that there is a TS error: https://github.com/metabase/metabase/actions/runs/7199526117/job/20089367211?pr=35863
>   ```
>   Error: frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/TypeSelector/TypeSelector.tsx(39,57): error 
>   TS2345: Argument of type '"menu" | ClickBehaviorType' is not assignable to parameter of type 'ClickBehaviorType'.
>     Type '"menu"' is not assignable to type 'ClickBehaviorType'.
>   ```

Hi, @kamilmielnik !

I rebased the branch with master yesterday and fixed the TypeScript error. Would love to hear your feedback. Thanks!

kamilmielnik commented on e2e/test/scenarios/dashboard-cards/click-behavior.cy.spec.js:
```diff
@@ -179,6 +179,13 @@ describe("scenarios > dashboard > dashboard cards > click behavior", () => {
       editDashboard();
 
       getDashboardCard().realHover().icon("click").should("not.exist");
+      // When the default menu is selected, it should've visual cue (metabase#34848)
+      cy.get("aside")
+        .findByText("Open the Metabase drill-through menu")
+        .parent()
+        .parent()
+        .should("have.attr", "aria-selected", "true")
+        .should("have.css", "background-color", "rgb(80, 158, 227)");
```
This assertion fails and should not be here.
Notice that line 181 makes sure that it's not possible to open the `aside` for that dashboard card.

```suggestion
```

ibedwi replied:
Hi @kamilmielnik . I apologize, this assertion might be moved to the wrong place when I'm rebasing from the `master` branch earlier this year. I moved the assertion to the correct place.

ibedwi committed:
fix: e2e test

ibedwi committed:
fix: remove set type selector after unset column

ibedwi committed:
fix: show type selector when click behavior is undefined

ibedwi committed:
fix: show type selector when finalClickBehavior is default (actionMenu)

kamilmielnik reviewed (approved):
Looks good!
